File: frontend/src/utils/controlResponse.ts
import { isObject, pickString } from '~/lib/jsonPick'

/** Extract the tool_name from a control_request payload */
export function getToolName(payload: Record<string, unknown>): string {
  const request = payload.request as Record<string, unknown> | undefined
  return (request?.tool_name as string) ?? ''
}

/** Extract the input from a control_request payload */
export function getToolInput(payload: Record<string, unknown>): Record<string, unknown> {
  const request = payload.request as Record<string, unknown> | undefined
  return (request?.input as Record<string, unknown>) ?? {}
}

/**
 * Builds a control_response JSON object that allows a tool use.
 */
export function buildAllowResponse(
  requestId: string,
  updatedInput: Record<string, unknown>,
  opts?: { permissionMode?: PermissionMode, clearContext?: boolean },
): Record<string, unknown> {
  return {
    type: 'control_response',
    permissionMode: opts?.permissionMode,
    ...(opts?.clearContext ? { clearContext: true } : {}),
    response: {
      subtype: 'success',
      request_id: requestId,
      response: {
        behavior: 'allow',
        updatedInput,
      },
    },
  }
}

/**
 * The placeholder reject message emitted when the user declines a control request WITHOUT
 * typing a reason. Must stay byte-identical to the backend's `ControlRejectedByUserMessage`
 * (backend/internal/worker/agent/factory.go), whose `NormalizeRejectionMessage` collapses it
 * to "" so a bare deny renders "Rejected" rather than leaking this placeholder as if it were
 * typed feedback. Kept as ONE frontend constant so the producer sites (buildDenyResponse's
 * default and any caller wanting a bare deny) can't drift from each other.
 */
export const CONTROL_REJECTED_BY_USER_MESSAGE = 'Rejected by user.'

/**
 * Builds a control_response JSON object that denies a tool use. Omit `message` for a bare
 * deny with no typed reason (it falls back to CONTROL_REJECTED_BY_USER_MESSAGE).
 */
export function buildDenyResponse(
  requestId: string,
  message?: string,
): Record<string, unknown> {
  return {
    type: 'control_response',
    response: {
      subtype: 'success',
      request_id: requestId,
      response: {
        behavior: 'deny',
        // Ensure the message is never empty — Claude Code SDK converts deny
        // responses into tool_result with is_error=true, and the Anthropic API
        // rejects empty content when is_error is set.
        message: message || CONTROL_REJECTED_BY_USER_MESSAGE,
      },
    },
  }
}

/**
 * Trim a control-response reject reason and collapse the
 * {@link CONTROL_REJECTED_BY_USER_MESSAGE} placeholder (the auto-filled "declined without a
 * reason" text) to "". Mirrors the backend's `NormalizeRejectionMessage` so a bare deny reads
 * as "Rejected" on both sides instead of leaking the placeholder as typed feedback.
 */
export function normalizeRejectionMessage(message: string): string {
  const trimmed = message.trim()
  return trimmed === CONTROL_REJECTED_BY_USER_MESSAGE ? '' : trimmed
}

/**
 * Decode the neutral approve/reject control-response envelope
 * (`{response:{request_id, response:{behavior, message}}}`) from an already-parsed OBJECT, the
 * shape `buildAllowResponse` / `buildDenyResponse` produce and the Claude/Codex-envelope native
 * response persists. Returns the trimmed request id, behavior, and rejection message (with the
 * {@link CONTROL_REJECTED_BY_USER_MESSAGE} sentinel collapsed to ""), or null when the object
 * doesn't carry a recognizable allow/deny behavior. The single home for reading this shape,
 * shared by the byte-oriented {@link decodeControlResponseBehavior} and the persisted-row
 * renderer (persistedControlResponse.ts).
 */
export function decodeControlBehaviorEnvelope(
  obj: unknown,
): { requestId: string, behavior: 'allow' | 'deny', message: string } | null {
  if (!isObject(obj))
    return null
  const outer = isObject(obj.response) ? obj.response : undefined
  const inner = isObject(outer?.response) ? outer.response as Record<string, unknown> : undefined
  const behavior = pickString(inner, 'behavior', '').trim()
  if (behavior !== 'allow' && behavior !== 'deny')
    return null
  return {
    requestId: pickString(outer, 'request_id', '').trim(),
    behavior,
    message: normalizeRejectionMessage(pickString(inner, 'message', '')),
  }
}

/**
 * Decode the inverse of `buildAllowResponse` / `buildDenyResponse`: a
 * `control_response` envelope's serialized bytes back into its `behavior`.
 * Returns null on any decode failure or when the envelope doesn't carry a
 * recognizable behavior — callers can then fall back to their own state.
 *
 * Providers that want to translate the shared AskUserQuestion
 * response into their own wire format (e.g. Pi's `extension_ui_response`
 * shape) call this helper instead of reaching into the nested
 * `response.response` themselves.
 */
export function decodeControlResponseBehavior(content: Uint8Array): 'allow' | 'deny' | null {
  let parsed: unknown
  try {
    parsed = JSON.parse(new TextDecoder().decode(content))
  }
  catch {
    return null
  }
  return decodeControlBehaviorEnvelope(parsed)?.behavior ?? null
}

/**
 * Builds a control_request JSON string for changing Claude Code's permission mode.
 * The hub detects this format and sends it as raw input to Claude Code's stdin.
 * Uses the same wire protocol as the Agent SDK's setPermissionMode().
 */
export type PermissionMode = string

File: backend/internal/worker/agent/control_response.go
package agent

import (
	"encoding/json"
	"strings"

	leapmuxv1 "github.com/leapmux/leapmux/generated/proto/leapmux/v1"
)

// ControlResponseContext is the pure provider input for interpreting a frontend
// control response. The service loads the stored control request once, extracts the
// shared metadata, and passes both payloads here; providers must not perform I/O.
type ControlResponseContext struct {
	RequestID       string
	RequestPayload  json.RawMessage
	ResponseContent []byte
	ToolName        string
	PlanApproval    *leapmuxv1.PlanApprovalSettings
}

// ControlResponseResolution is the provider-owned interpretation of a control
// response. The service executes side effects from this plan: it persists the
// structured control-response row, deletes control requests, mutates plan-mode
// settings, and forwards Content.
type ControlResponseResolution struct {
	Content []byte
	// Feedback must use a separate user input when the native response cannot carry it safely.
	Feedback      string
	SelfDisplayed bool
	// Withhold prevents forwarding a response that the provider cannot read.
	// Withhold covers that case alone. The service returns an error and keeps the
	// request available for another answer.
	//
	// The service refuses an answer whose control request row is gone, and that
	// refusal runs before the service reads Withhold. A withheld resolution
	// therefore always describes a request that still exists.
	//
	// Empty Content cannot express this decision because the service restores the original response bytes.
	Withhold        bool
	PlanModeControl PlanModeControlKind
}

func DefaultControlResponseResolution(ctx ControlResponseContext) ControlResponseResolution {
	return restoreControlResponseID(ctx)
}

// defaultControlResponseRequestID reads the stored-control-request lookup id from a raw frontend
// control response. It tries the neutral approve/reject envelope first
// (DecodeControlBehavior: {response:{request_id, ...}}, which every provider's frontend emits) and
// falls back to a top-level JSON-RPC id (ExtractJSONRPCID, used by the ACP family and Codex). When
// a payload carries BOTH -- a top-level id AND a nested response.request_id -- the nested envelope
// id wins, because the pending control_request row is keyed by response.request_id. Returns "" when
// neither shape yields a non-empty id. Shared by every provider's ControlResponseRequestID; no
// provider narrows it, since narrowing to one shape would drop the other's real flows.
//
// The JSON-RPC branch returns the id TEXT, and it must not canonicalize it. The browser echoes
// the worker's own request id -- the "jsonrpc:"-prefixed key PublishControlRequest stored the row
// under -- as a JSON STRING in that field (buildJsonRpcResult), because a JSON number loses
// precision there. So the text already IS the stored key, and running it through
// NewControlRequestIdentity would prefix it a second time and match nothing.
func defaultControlResponseRequestID(content []byte) string {
	if requestID, _, _, ok := DecodeControlBehavior(content); ok && requestID != "" {
		return requestID
	}
	if _, requestID, ok := ExtractJSONRPCID(content); ok {
		return requestID
	}
	return ""
}

func (ProviderDefaults) ControlResponseRequestID(content []byte) string {
	return defaultControlResponseRequestID(content)
}

func (ProviderDefaults) ResolveControlResponse(ctx ControlResponseContext) ControlResponseResolution {
	return DefaultControlResponseResolution(ctx)
}

// ControlBehaviorEnvelope describes the shared approval and rejection fields.
// The decoder and plan service use this type. LeapMux plan settings remain outside provider JSON.
type ControlBehaviorEnvelope struct {
	Response struct {
		RequestID string `json:"request_id"`
		Response  struct {
			Behavior string `json:"behavior"`
			Message  string `json:"message"`
			// Choice is the id of one of the choices a control request offered
			// beside its plain approve and reject: a plan's own approach, a
			// refusal that also ends plan mode. Empty for a plain decision. The
			// browser's plan control writes it (withControlChoice in
			// frontend/src/utils/controlResponse.ts), and each provider maps the
			// id onto its own wire answer.
			Choice string `json:"choice"`
		} `json:"response"`
	} `json:"response"`
}

// DecodeControlChoice returns the trimmed choice a control response carries
// beside its behavior, or "" for a plain decision and for bytes that are not
// JSON. See ControlBehaviorEnvelope.
func DecodeControlChoice(content []byte) string {
	var cr ControlBehaviorEnvelope
	if err := json.Unmarshal(content, &cr); err != nil {
		return ""
	}
	return strings.TrimSpace(cr.Response.Response.Choice)
}

// DecodeControlBehavior decodes the frontend's neutral approve/reject control-response envelope
// (ControlBehaviorEnvelope), returning the trimmed request id, behavior, and rejection
// message. ok is false only when the bytes don't parse as JSON. The message is the user's typed
// rejection reason, with the ControlRejectedByUserMessage sentinel (an auto-filled placeholder,
// not a real reason) collapsed to "". The SINGLE home for the sentinel rule, shared by the Codex
// feedback path, the Cursor create-plan transform, and the shared control-response request-id
// default (defaultControlResponseRequestID), so a sentinel change lands in exactly one place.
func DecodeControlBehavior(content []byte) (requestID, behavior, message string, ok bool) {
	var cr ControlBehaviorEnvelope
	if err := json.Unmarshal(content, &cr); err != nil {
		return "", "", "", false
	}
	requestID = strings.TrimSpace(cr.Response.RequestID)
	behavior = strings.TrimSpace(cr.Response.Response.Behavior)
	message = NormalizeRejectionMessage(cr.Response.Response.Message)
	return requestID, behavior, message, true
}

// NormalizeRejectionMessage trims a control-response reject reason and collapses the
// ControlRejectedByUserMessage placeholder (the auto-filled "declined without a reason" text) to
// "". The SINGLE home for the deny-feedback rule, shared by DecodeControlBehavior (which decodes
// the raw wire bytes) and the service's controlResponsePlan.rejectionMessage accessor (which reads
// the already-decoded plan), so a sentinel-value change lands in exactly one place instead of two.
func NormalizeRejectionMessage(message string) string {
	message = strings.TrimSpace(message)
	if message == ControlRejectedByUserMessage {
		return ""
	}
	return message
}

// Control response behavior values (shared protocol between frontend and backend).
const (
	ControlBehaviorAllow = "allow"
	ControlBehaviorDeny  = "deny"
)

// ControlRejectedByUserMessage is the placeholder reject message the frontend emits when
// the user declines a control request WITHOUT typing a reason (buildDenyResponse in
// frontend utils/controlResponse.ts). The backend treats it as "no feedback" -- it is not
// shown as the user's answer -- so every deny-with-feedback path compares against this one
// constant instead of re-spelling the literal (which must stay in lockstep with the
// frontend producer).
const ControlRejectedByUserMessage = "Rejected by user."
